src/loaders/workspace_downloader/workspace_downloader.py

```diff
@@ -280,29 +295,173 @@ def process_input(conf):
+        upa_dir = os.path.join(conf.pth, upa)
+        metafile = os.path.join(upa_dir, f"{upa}.meta")
+        if not os.path.isdir(upa_dir) or not loader_helper.is_upa_info_complete(upa_dir):
+
+            # remove legacy upa_dir to avoid FileExistsError in hard link
+            if os.path.isdir(upa_dir):
+                shutil.rmtree(upa_dir)
+
+            # cfn points to the assembly file outside of the container
+            # get_assembly_as_fasta writes the file to /kb/module/workdir/tmp/<filename> inside the container.
+            # workdir is shared between the container and the external file system
+            # Any file path get_assembly_as_fasta returns will be relative to inside the container, and so is not useful for this script
+            cfn = os.path.join(conf.job_dir, "workdir/tmp", upa)
+            # upa file is downloaded to cfn
+            conf.asu.get_assembly_as_fasta({"ref": upa.replace("_", "/"), "filename": upa})
+
+            # each upa in output_dir as a separate directory
+            os.makedirs(upa_dir, exist_ok=True)
+
+            dst = os.path.join(upa_dir, f"{upa}.fa")
+            # Hard link .fa file from job_dir to output_dir in WS
+            os.link(cfn, dst)
+
+            # save meta file with relevant object_info
+            with open(metafile, "w", encoding="utf8") as json_file:
+                json.dump(_process_object_info(obj_info, genome_upa), json_file, indent=2)
+
+            print("Completed %s" % (upa))
+        else:
+            print(f"Skip downloading {upa} as it already exists")
+
+        if conf.retrieve_sample:
+            _download_sample_data(conf, genome_upa.replace("/", "_"), metafile)
+
+
+def _download_sample_data(conf: Conf, upa: str, metafile: str) -> None:
+    # retrieve sample data from sample service and save to file
+    # additionally, retrieve node data from the sample data and save it to a file
+    # NOTE: upa is in the format of "A_B_C"
+
+    with open(metafile, "r", encoding="utf8") as json_file:
+        meta = json.load(json_file)
+
+    upa_dir = Path(metafile).parent
+    sample_file_name = f"{upa}.{loader_common_names.SAMPLE_FILE_EXT}"
+    sample_file = os.path.join(upa_dir, sample_file_name)
+    if _check_file_exists(loader_common_names.SAMPLE_FILE_KEY, meta, sample_file):
+        print(f"Skip downloading sample for {upa} as it already exists")
+        return
+
+    sample_ret = _retrieve_sample(conf, upa)
+    if not sample_ret:
+        return
+
+    with open(sample_file, "w", encoding="utf8") as json_file:
+        json.dump(sample_ret, json_file, indent=2)
+
+    sample_prepared_name = f"{upa}.{loader_common_names.SAMPLE_PREPARED_EXT}"
+    sample_prepared_file = os.path.join(upa_dir, sample_prepared_name)
+    if _check_file_exists(loader_common_names.SAMPLE_PREPARED_KEY, meta, sample_prepared_file):
+        print(f"Skip generating sample node tree for {upa} as it already exists")
+        return
```

Comment:
Hmm, this means that somehow there was no sample file but there was a prepared file, which should be impossible and indicates something weird has happened. I think what I would do here is log the anomaly and overwrite the existing prepared file

Reply:
or throw an error, that'd also work

Reply:
hmmm did you mis-read `loader_common_names.SAMPLE_PREPARED_KEY`? `SAMPLE_FILE_KEY` <--> `SAMPLE_PREPARED_KEY`.

Reply:
I don't think so. UUIC the code
* checks to see if there's a sample file. If there is, it `returns`
* creates the sample file
* checks to see if there's a prepared file
     * if there is, it returns, but that means that
         * the "raw" sample file didn't exist at the start of the method
         * the "prepared" sample file did exist, which should be impossible

Right? Or am I missing something?

Reply:
you are right. updated logic should prevent this happening.

Reply:
Hmm, thinking through how this should work here's what I would do:

* check the metadata file for the 3 sample keys. If they're both there, `return`.
* Pull the sample data
* Save both the files no matter what
* Open the metadata file again to try to avoid race conditions (we should do this in a DB later)
* Add the sample keys and save.

This would mean
* We don't pull the sample when we already have the data
* There's no chance of mixing data between 2 different samples if the sample for an object gets relinked

Reply:
I pulled it every time because I think ws can link to different samples?

Reply:
by the way, once we download samples the first time, how likely will we ever rerun it again?

Reply:
> I pulled it every time because I think ws can link to different samples?

It's a possibility, but for our current use case, not very likely. If we wanted to check that I'd just add a force toggle for sample downloads

Reply:
> by the way, once we download samples the first time, how likely will we ever rerun it again?

For our current use case pretty unlikely I'd guess, but possibly if we find errors in the sample, that kind of thing

Reply:
👍

Reply:
:+1: